weave/trace_server/client_server_common/digest_builder.py

```diff
@@ -0,0 +1,195 @@
+    stabilized = _stabilize_refs_in_json_like(val)
+    canonical = _canonicalize_json_like(stabilized)
+    return _json_digest(canonical)
+
+
+def table_digest_from_row_digests(row_digests: list[str]) -> str:
+    """Generate a stable digest for a table from row digests.
+
+    The resulting digest is **order-dependent**: `[row1, row2]` hashes differently than
+    `[row2, row1]`.
+
+    Args:
+        row_digests (list[str]): Row digests in the intended table order.
+
+    Returns:
+        str: Hex-encoded SHA-256 digest.
+    """
+    table_hasher = hashlib.sha256()
+    for row_digest in row_digests:
+        table_hasher.update(row_digest.encode())
+    return table_hasher.hexdigest()
+
+
+def _json_digest(json_val: Any) -> str:
+    # IMPORTANT: This must stay stable over time. We explicitly sort dict keys so that
+    # semantically-identical JSON objects (that differ only in insertion order) hash the same.
+    #
+    # NOTE: We intentionally do *not* change json.dumps separators/whitespace settings here,
+    # because that would change all digests. Keep encoding defaults unless there's a strong
+    # reason to change them (and a migration plan).
+    return _str_digest(json.dumps(json_val, sort_keys=True))
+
+
+E = TypeVar("E")
+
+
+def _map_val(obj: E, func: Callable[[E], E]) -> E:
+    if isinstance(obj, dict):
+        return cast(E, {k: _map_val(v, func) for k, v in obj.items()})
+    if isinstance(obj, list):
+        return cast(E, [_map_val(v, func) for v in obj])
+    if isinstance(obj, tuple):
+        return cast(E, tuple(_map_val(v, func) for v in obj))
+    if isinstance(obj, set):
+        # Sets are not supported anywhere in safe_digest; error early and avoid iterating
+        # over an unordered collection.
+        raise TypeError(
+            "Cannot hash set values in digest_builder; convert to a list/tuple first."
+        )
+    return func(obj)
+
+
+def _canonicalize_json_like(obj: Any) -> Any:
+    """Canonicalize JSON-like objects for stable hashing.
+
+    This is intentionally ref-unaware: it only deals with deterministic encoding of
+    containers (dict/list/tuple/set) so that stable hashing is possible.
+    """
+    if isinstance(obj, dict):
+        return {k: _canonicalize_json_like(obj[k]) for k in sorted(obj.keys())}
+    if isinstance(obj, list):
+        return [_canonicalize_json_like(v) for v in obj]
+    if isinstance(obj, tuple):
+        return [_canonicalize_json_like(v) for v in obj]
+    if isinstance(obj, set):
+        # Sets are inherently unordered; hashing them would be sensitive to unstable
+        # iteration order. Require callers to convert sets to an ordered type first.
+        raise TypeError(
+            "Cannot hash set values in digest_builder; convert to a list/tuple first."
+        )
+    return obj
+
+
+def _stabilize_refs_in_json_like(val: Any) -> Any:
+    """Return a JSON-like value with Weave ref strings normalized."""
+
+    def mapper(v: Any) -> Any:
```

Comment:
This is not a closure, so it's better to put this at the top level.  Otherwise you are redefining this def (expensive) in a likely hot path of the code